File: data_structure/Tree/617_Merge_Two_Binary_Trees.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
############################################################################
#Author:polya 
#Email:polyaluthor@gmail.com
#Last updated:2019.10.22
#Version: 1.0 first
#need do: unit test
#Purpose: 617_Merge_Two_Binary_Trees.py
#Main logic principle:
#
#
#Usage:
#     example: python 617_Merge_Two_Binary_Trees.py
############################################################################
"""
import os
import sys
import datetime
import argparse
import pandas as pd
#
#
#Get Argv
#get ready parameters and out help inforamtion
#############################################################################
def get_args():
	parse=argparse.ArgumentParser(description='annot expression in one file')
	parse.add_argument('-u','--usage',help="usage: python $0  -in 1,2,3 ")
	parse.add_argument('-in','--in_list',type=str,help="input ready files ")
	parse.add_argument('-out','--out_file',type=str,help="set out name of files: eg.select.txt ")
	args=parse.parse_args()
	return vars(args)
############################################################################
#
#
#Do somthing for data
class TreeNode(object):
	def __init__(self,key=None,lchild=None,rchild=None):
		self.key=key
		self.lchild=lchild
		self.rchild=rchild
############################################################################
class Solution:
	def __init__(self,data_list):
		self.it=iter(data_list)
	def createBiTree(self,bt=None):
		try:
			next_data=next(self.it)
			if next_data is '#':
				bt=None
			else:
				bt=TreeNode(next_data)
				bt.lchild=self.createBiTree(bt.lchild)
				bt.rchild=self.createBiTree(bt.rchild)
		except Exception as e:
			print(e)
		return bt
		#print(self.k)
	def preOrderTrave(self,bt):
			if bt is not None:
				print(bt.key,end=" ")
				self.preOrderTrave(bt.lchild)
				self.preOrderTrave(bt.rchild)
	def printTrave(self,bt):
			print("\n先序遍历: ", end="")
			self.preOrderTrave(bt)
	def mergeTrees(self,t1,t2):
		if  t1==None and t2==None:
			return None
		elif t1==None:
			return t2
		elif t2==None:
			return t1
		else:
			t1.key+=t2.key
			t1.left=self.mergeTrees(t1.lchild,t2.lchild)
			t1.right=self.mergeTrees(t1.rchild,t2.rchild)
		return t1
#
#
#Call steps for ends
############################################################################
def call_main_steps():
	args=get_args()
	start =datetime.datetime.now()
	#arr=[1,12,-5,-6,50,3,100,20,11]
	#arr=[9,2,1,2,1,9]
	arr=[1,1,3,5,6,6,8]
	arr1=[1,1,3,5,6,6,7]
#self test##################################
	#data = input("Please input the node value: ")
	data_list = list(arr1)
	btree = Solution(data_list)
	btree1=Solution(arr)
	root=btree.createBiTree()
	root1=btree1.createBiTree()
	btree.printTrave(root)
	btree.printTrave(root1)
	btr = Solution(data_list)
	rootMe=btr.mergeTrees(root,root1)
	print("\ntype of rootMe\n"+str(type(rootMe)))
	btree.printTrave(rootMe)
	print("\n\n============================================================")
	print("Hello polya,welcome to the programming world")
	end=datetime.datetime.now()
	#print("The programming running time : "+str(end-start))
	print('The programming running time: %s Seconds'%(end-start))
	print("The programming finished time : "+str(end))
	print("============================================================")
#input from out############################
#unit test ################################ 
#
#main
############################################################################
if __name__=='__main__':
	call_main_steps()

File: data_structure/Tree/617_Merge_Two_Binary_Trees.cpp
/*############################################################################
#Author:polya 
#Emalil:polyaluthor@gmail.com
#617_Merge_Two_Binary_Trees.cpp
#Last updated:2019.10.23
#history version
#Version: 0.1
#Purpose:
#Main logic principle:
#Usage:
#     example: model.cpp
g++ model.cpp -o model -std=gnu++11
make model
./model
g++ 617_Merge_Two_Binary_Trees.cpp -o 617 -std=gnu++11
make 617
./617
############################################################################*/
//logic outline:print information
//include input and out library
//logic struture:
//
#include <iostream>
#include <ctime>
#include <vector>
#include <queue>
#include <iterator>
#include <sstream>
#include <algorithm>
#include "stdio.h"
using namespace std;
//define a binary tree node
struct TreeNode{
		int val;
		TreeNode *left;
		TreeNode *right;
		TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
//define class for solution
class Solution {
	public:
		TreeNode* mergeTrees(TreeNode* t1,TreeNode* t2){
			if(!t1){
				return t2;
			}
			else if(!t2){
				return t1;
			}else{
				TreeNode* node = new TreeNode(t1->val + t2->val);
				node->left = mergeTrees(t1->left, t2->left);
				node->right = mergeTrees(t1->right, t2->right);
			return node;
			}
		}
// travel binary tree
// flag==0 preOrder  flag=1 inOrder flag=2 postOrder
	void displayInfo_Order(TreeNode* t,vector<int> &result,int flag){
	if(flag==0){
		if(t!=NULL){
			result.push_back(t->val);
			displayInfo_Order(t->left,result,0);
			displayInfo_Order(t->right,result,0);
		}
	}else if(flag==1){
		if(t!=NULL){
			displayInfo_Order(t->left,result,1);
			result.push_back(t->val);
			displayInfo_Order(t->right,result,1);
		}
	}else{
		if(t!=NULL){
			displayInfo_Order(t->left,result,2);
			displayInfo_Order(t->right,result,2);
			result.push_back(t->val);
		}
	}
	}
	//travel BT
	void traverse(vector<int> nums){
		vector<int>::size_type size = nums.size();
		for (size_t i = 0; i < size; i++){
			cout << nums[i] << ' ';
		}
		cout << endl;
	}
	//init BT
	TreeNode* initBTree(int elements[], int size){
		if (size < 1){
			return NULL;
		}
		TreeNode **nodes = new TreeNode*[size];
		for (int i = 0; i < size; i++){
		/*/v1
			if (elements[i] == 0){
				nodes[i] = NULL;
			}else{
				nodes[i] = new TreeNode(elements[i]);
			}
		*/
			nodes[i] = new TreeNode(elements[i]);
		}
		queue<TreeNode*> nodeQueue;
		nodeQueue.push(nodes[0]);
		TreeNode *node;
		int index = 1;
		while (index < size){
			node = nodeQueue.front();
			nodeQueue.pop();
			nodeQueue.push(nodes[index++]);
			node->left = nodeQueue.back();
			nodeQueue.push(nodes[index++]);
			node->right = nodeQueue.back();
		}
		return nodes[0];
	}
};
//call main function
int main(int argc, char* argv[])
{
//self test ###################################################################
	clock_t begin = clock();
	cout<<"Hello polya,welcome to the programming world" <<endl;
	cout<<"===========================================" <<endl;
	cout<<"//out element: " <<endl;
//run programming############################
	int nums[] = { 1, 2, 3, 0, 1, 5, 6, 10, 7 };
	int nums_2[] = { 1, 2, 3, 4, 9, 5, 6, 10, 7 };
	Solution bT;
	//init BTree
	TreeNode *bt = bT.initBTree(nums, 9);
	TreeNode *bt2 = bT.initBTree(nums_2, 9);
	//display
	vector<int> preResult;
	bT.displayInfo_Order(bt,preResult,1);
	bT.traverse(preResult);
	preResult.clear();
	bT.displayInfo_Order(bt2,preResult,1);
	bT.traverse(preResult);
	//merge
	vector<int> mergeResult;
	TreeNode *btM=bT.mergeTrees(bt,bt2);
	bT.displayInfo_Order(btM,mergeResult,1);
	bT.traverse(mergeResult);
	cout<<" The out One "<<endl;
//###########################################
	cout<<"===========================================" <<endl;
	clock_t end = clock();
	double elapsed_secs = static_cast<double>(end - begin) / CLOCKS_PER_SEC;
	time_t now = time(0);
	char* dt = ctime(&now);
	cout<<"The programming had finished at time : "<< dt << "==============================" <<endl;
	cout <<"The programming running time: "<< elapsed_secs << " s" << endl;
//out test ####################################################################
//unit test ###################################################################
	return 0;
}
